code/go/0chain.net/core/common/rate_limiter.go

```diff
@@ -62,11 +79,87 @@ func UserRateLimit(handler ReqRespHandlerf) ReqRespHandlerf {
 	}
 }
 
-//UserRateLimit - rate limiting for end user handlers
+//UseUserRateLimit - rate limiting for end user handlers
 func UseUserRateLimit(h http.Handler) http.Handler {
 	if !userRateLimit.RateLimit {
 		return h
 	}
 
 	return tollbooth.LimitFuncHandler(userRateLimit.Limiter, h.ServeHTTP)
 }
+
+// FileRateLimit is a custom of `tollbooth.LimitHandlerFunc` that uses custom values to build keys for limiting file actions.
+func FileRateLimit(handler ReqRespHandlerf) ReqRespHandlerf {
+	if fileRateLimit == nil || !fileRateLimit.RateLimit {
+		return handler
+	}
+
+	lmt := fileRateLimit.Limiter
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		keys := FileLimitBuildKeys(lmt, r)
+		httpError := tollbooth.LimitByKeys(lmt, keys)
+		if httpError != nil {
+			lmt.ExecOnLimitReached(w, r)
+			if lmt.GetOverrideDefaultResponseWriter() {
+				return
+			}
+			w.Header().Add("Content-Type", lmt.GetMessageContentType())
+			w.WriteHeader(httpError.StatusCode)
+			_, _ = w.Write([]byte(httpError.Message))
+			return
+		}
+
+		// There's no rate-limit error, serve the next handler.
+		handler(w, r)
+	}
+}
+
+// FileLimitBuildKeys return build keys to be used for file rate limiter.
+func FileLimitBuildKeys(lmt *limiter.Limiter, r *http.Request) []string {
+	baseKeys := tollbooth.BuildKeys(lmt, r)
+
+	client := r.Header.Get(ClientHeader)
+	baseKeys[0] = append(baseKeys[0], strings.ToLower(client))
+
+	// add connection_id when present (exists for upload)
+	conn, ok := requestField(r, "connection_id")
+	if ok {
+		baseKeys[0] = append(baseKeys[0], strings.ToLower(conn))
+		return baseKeys[0]
+	}
+
+	// OR add path hash (may exists for download)
+	pathHash := r.Header.Get("X-Path-Hash")
+	if pathHash != "" {
+		baseKeys[0] = append(baseKeys[0], strings.ToLower(pathHash))
+		return baseKeys[0]
+	}
+
+	// OR add path (may exists for download)
+	path := r.Header.Get("X-Path")
+	if path != "" {
+		baseKeys[0] = append(baseKeys[0], path)
+		return baseKeys[0]
+	}
+
+	// OR just empty
+	baseKeys[0] = append(baseKeys[0], "")
+	return baseKeys[0]
+}
+
+func requestField(r *http.Request, key string) (string, bool) {
+	if r.Form != nil {
```

Comment:
We can get URL query data from r.Form as well, so don't have to call the r.URL.Query below, and the code could be like:

```go
if err := r.ParseForm(); err != nil {
  return "", err
}

return r.Form.Get(key), nil
```